source/portfolio/portfolio_mad.cpp

```diff
@@ -0,0 +1,84 @@
+//
+// Created by eduardo on 15/04/2021.
+//
+
+#include "portfolio_mad.h"
+#include <iostream>
+#include <numeric>
+#include <random>
+namespace portfolio {
+
+    portfolio_mad::portfolio_mad(market_data_mad &mad_data) {
+        static std::default_random_engine generator =
+            std::default_random_engine(
+                std::chrono::system_clock::now().time_since_epoch().count());
+        std::uniform_real_distribution<double> d(0.0, 1.0);
+        for (auto a = mad_data.assets_begin(); a != mad_data.assets_end();
+             ++a) {
+            if (0.5 > d(generator)) { // asset is selected or not
+                assets_[*a] = d(generator);
+            } else {
+                assets_[*a] = 0.0;
+            }
+        }
+        normalize_allocation();
+        evaluate(mad_data);
+    }
+    void portfolio_mad::normalize_allocation() {
+        double total = total_allocation();
+        if (total == 1.0) {
```

Comment:
Por questões de aproximação, não podemos comparar números de ponto flutuante com `==`. O operador existe apenas para casos muito raros, onde temos garantia que o valor é exatamente `1.0` na conversão em binário. 

Para todos os outros casos, onde o número veio de outra função, isso quase sempre vai dar falso pois existem infinitos números reais. Mesmo que seja bem próximo de `1.0`, o seu `total` sempre vai ser algo como `0.99999999` ou `1.000000001`. Vários compiladores emitem erro em modo pedante para todas as comparações `==` em número real.

A solução é sempre definir um intervalo de erro [epsilon](https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon) aceitável para sua aplicação e usar `<` e `>` para ver se o número está dentro do seu intervalo.

Se for precisar destas comparações frequentemente, você pode colocar uma função `almost_equal` nos seus algoritmos e ir adaptando na medida que precisar:

```cpp
template<class T>
typename std::enable_if<!std::numeric_limits<T>::is_integer, bool>::type
    almost_equal(T x, T y, int precision = 2)
{
    // the machine epsilon has to be scaled to the magnitude of the values used
    // and multiplied by the desired precision in ULPs (units in the last place)
    return std::fabs(x-y) <= std::numeric_limits<T>::epsilon() * std::fabs(x+y) * precision
        // unless the result is subnormal
        || std::fabs(x-y) < std::numeric_limits<T>::min();
}
```

Coloquei uma precisão mínima como padrão aí mas pode experimentar ajustar como achar melhor na sua aplicação.

Reply:
Adicionei a função almost_equal conforme orientação.